cpp/src/arrow/record_batch.cc

```diff
@@ -37,6 +37,26 @@
 
 namespace arrow {
 
+namespace {
+// If there will only be one slice returned it is cheaper to just return the original
+// RecordBatch (no overhead from vector and std::shared_ptr copying on underlying arrays).
```

Comment:
left over comment from when I was going to down the rabbit whole.  I've moved all this code into the writer.cc to avoid debating the best way to expose this publicly.